File: R/progress.R
makeProgress <- function(end, name, freq, namelength = nchar(name), maxwait = 10) {
    nameMod <- unlist(strsplit(name, split = ''))
    nameMod <- nameMod[seq_len(namelength)]
    nameMod[is.na(nameMod)] <- ' '
    nameMod <- paste(nameMod, collapse = '')
    if (end/freq < 50) {
        freq <- floor(end/50)
    }
    prog <- list(
        end = end,
        name = nameMod,
        freq = freq,
        prog = 0,
        last = 0,
        maxwait = maxwait,
        lasttime = as.numeric(Sys.time())
    )
    createBar(prog)
    list2env(prog, parent = emptyenv())
}

progress <- function(progress) {
    progress$prog <- progress$prog + 1
    progress$last <- progress$last + 1
    currentTime <- as.numeric(Sys.time())
    if (progress$end == progress$prog) {
        createBar(progress)
    } else if (progress$last > progress$freq || currentTime - progress$lasttime > progress$maxwait) {
        progress$lasttime <- currentTime
        progress$last <- 0
        createBar(progress)
    }
    flush.console()
}
createBar <- function(progress) {
    p <- round(progress$prog / progress$end * 50)
    whites <- paste(rep(' ', nchar(progress$end) - nchar(progress$prog)), collapse = '')
    pString <- paste(c(rep('=', p), rep(' ', 50 - p)), collapse = '')
    pString <- paste0(progress$name, ' |', pString, '| ')
    pString <- paste0('\r', pString, whites, progress$prog, '/', progress$end)
    cat(pString)
    flush.console()
}

File: src/progress.cpp
#include <time.h>
#include <Rcpp.h>
#include "progress.h"

using namespace Rcpp;

int nDigits(int n) {
    int number_of_digits = 0;
    do {
        ++number_of_digits; 
        n /= 10;
    } while (n);
    return number_of_digits;
}

Progress::Progress (int e, std::string n, bool s) {
    show = s;
    end = e;
    name = n;
    freq = e / 50;
    prog = 0;
    last = 0;
    time(&lasttime);
    maxwait = 10;
}
Progress::Progress (int e, std::string n, int f, bool s) {
    show = s;
    end = e;
    name = n;
    if (e/f < 50) {
        freq = e / 50;
    } else {
        freq = f;
    }
    prog = 0;
    last = 0;
    time(&lasttime);
    maxwait = 10;
}
void Progress::createBar() {
    if (show) {
        int p = 50 * double(prog) / end;
        if (p > 50) p = 50;
        int countWhite = nDigits(end) - nDigits(prog);
        Rcout << "\r" << name << " |" << std::string(p, '=') << std::string(50-p, ' ') << "| " << std::string(countWhite, ' ') << prog << "/" << end << std::flush;
    }
}
void Progress::start() {
    createBar();
}
void Progress::finish() {
    prog = end;
    createBar();
}
void Progress::increment() {
    ++prog;
    if (prog > end) prog = end;
    ++last;
    time_t timer;
    time(&timer);
    
    if (end == prog) {
        createBar();
    } else if (last >= freq || maxwait < difftime(timer, lasttime)) {
        R_CheckUserInterrupt();
        lasttime = timer;
        last = 0;
        createBar();
    }
}